stdlib/public/core/MutableCollection.swift

```diff
@@ -182,16 +182,32 @@ where SubSequence: MutableCollection
     _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
   ) rethrows -> R?
 
-  /// Call `body(p)`, where `p` is a pointer to the collection's
-  /// mutable contiguous storage.  If no such storage exists, it is
-  /// first created.  If the collection does not support an internal
-  /// representation in a form of mutable contiguous storage, `body` is not
-  /// called and `nil` is returned.
+  /// Executes a closure on the collection's contiguous storage.
+  ///
+  /// This method calls `body(p)`, where `p` is a pointer to the collection’s
+  /// contiguous storage. If no such storage exists, the collection creates it.
+  /// If the collection doesn’t support an internal representation in a form of
+  /// contiguous storage, the method doesn’t call `body` and returns `nil`.
   ///
   /// Often, the optimizer can eliminate bounds- and uniqueness-checks
-  /// within an algorithm, but when that fails, invoking the
-  /// same algorithm on `body`\ 's argument lets you trade safety for
-  /// speed.
+  /// within an algorithm, but when that fails, invoking the same algorithm on
+  /// `body`'s argument lets you trade safety for speed.
+  ///
+  /// Successive calls to this method don't necessarily provide the same
+  /// pointer every time.
+  ///
+  /// This method makes no guarantees about the state of the collection if
+  /// the closure throws an error. Any changes made before the throw may
+  /// or may not be reflected in the collection. This is because the closure
+  /// could receive direct access to the collection's internal storage, or a
+  /// temporary copy. Therefore, the closure should always perform any
+  /// necessary cleanup.
+  ///
+  /// - Parameter body: A closure that receives an in-out
+  /// `UnsafeMutableBufferPointer` to the collection's contiguous storage.
```

Comment:
The UMBP is `inout` in order to enable ready use of `MutableCollection` API that require `self` to be mutable. In all implementations in the stdlib, swapping out the UMBP is an error that leads to a crash. Given that, mentioning the in-out-ness of the argument simply shines a light on a confusing aspect without a necessary warning.

Reply:
What would you think of adding another sentence of guidance to that effect, something like:

> Use the buffer pointer directly; do not replace it, as this leads to a crash in all implementations of this method within the standard library.

Reply:
I actually just ran into this myself and had to be corrected by my team, so mentioning that does seem good!

Reply:
Added a line about not replacing the UMBP.